### PR TITLE
feat(create-waku): install dependencies automatically when creating a new waku project

### PR DESCRIPTION
**Pull request by ojj1123**
follow up https://github.com/dai-shi/waku/pull/728

- [x] implement a new feature

### CODE REVIEW

**Comment by ojj1123** [comment]
스크린샷 2024-07-29 00 04 45

Some tests fail. I think that some tests of `create-waku.spec` is outdated due to some new features developed.
I try to add the test in this PR. But I will be busy next week, so if I'm done with my work, then I will work this.

**Comment by ojj1123** [comment]
[bb908ad](https://github.com/dai-shi/waku/pull/808/commits/bb908adc277a4680ba9e3900bfa9ed1a6b21411f)

I did comment all of changes. All tests pass. From now, I will see what change lead to fail.

File: packages/create-waku/src/index.ts [hunk 1]
```diff
@@ -202,15 +203,37 @@ async function init() {
...
   // 2. and then install dependencies
-
   console.log(`\nDone. Now run:\n`);
   console.log(`${bold(green(`cd ${targetDir}`))}`);
   console.log(`${bold(green(commands.install))}`);
   console.log(`${bold(green(commands.dev))}`);
   console.log();
+  // console.log(`Installing dependencies by running ${commands.install}...`);
+
+  // process.chdir(targetDir);
```
Comment (dai-shi) [inline packages/create-waku/src/index.ts:215]: My guess is that this makes it fail.
```suggestion
  process.chdir(targetDir);
```

**Comment by ojj1123** [comment]
[f1a7a6d](https://github.com/dai-shi/waku/pull/808/commits/f1a7a6d3e591e327b0e622959af713e0338234b9)

Using `spawn`, the tests fail on windows.

**Comment by dai-shi** [comment]
Ah, it's a windows only issue...

**Comment by ojj1123** [comment]
I feel like it can be solved if using [cross-spawn](https://www.npmjs.com/package/cross-spawn), but I didn't try to do that yet.

You were like that you don't want to use the external lib, so I try to solve it w/o that lib!

**Comment by dai-shi** [comment]
Yes, it's correct that I don't generally want to use external libraries.
I hope you can work it around, but otherwise, using `cross-spawn` might be acceptable if that's the only reasonable way.

**Comment by ojj1123** [comment]
https://github.com/dai-shi/waku/pull/808/commits/33affcc33fc91ed36b2cb25b94dfd8c2ebf11279

As I added shell option of `spawn` on only windows, the tests passed. I don't know why it is working well, b[ut `shell` option of `spawn` is `false` by default.](https://nodejs.org/api/child_process.html#child_processspawncommand-args-options).
I found the workaround having something to do with this issue.
👉 https://github.com/nodejs/node/issues/52681#issuecomment-2076426887

**Comment by dai-shi** [comment]
https://nodejs.org/docs/latest/api/child_process.html#child_processspawncommand-args-options

Sounds good!

**Comment by ojj1123** [comment]
I don't know why tests fail. I added shell option to `true` if being on `win32`. So I'm gonna try to take cross-spawn and then check the test status.

File: e2e/create-waku.spec.ts [hunk 2]
```diff
@@ -43,7 +43,10 @@ test('should create waku with default setup work', async () => {
   expect(files).toContain('package.json');
   expect(files).toContain('src');
   expect(files).toContain('tsconfig.json');
-  await terminate(childProcess.pid!);
+  exec(`rm -rf ${cwd}`);
+  // Why is this function needed?
+  // Can not terminate sub process automatically when tests are finished?
+  // await terminate(childProcess.pid);
```
Comment (ojj1123) [inline e2e/create-waku.spec.ts:49]: @himself65

Could you explain why `terminate` is needed?

Comment (ojj1123) [reply e2e/create-waku.spec.ts:49]: If adding this line, the error occurs below:

```
[[chromium] › create-waku.spec.ts:8:1 › should create waku with default setup work: e2e/create-waku.spec.ts#L1](https://github.com/dai-shi/waku/pull/808/files#annotation_26269905744)
  1) [chromium] › create-waku.spec.ts:8:1 › should create waku with default setup work ─────────────
    Error: kill ESRCH

        at handlePsTreeCallback (/Users/runner/work/waku/waku/node_modules/.pnpm/terminate@2.8.0/node_modules/terminate/handlePsTreeCallback.js:28:13)
        at /Users/runner/work/waku/waku/node_modules/.pnpm/terminate@2.8.0/node_modules/terminate/terminate.js:44:5
        at /Users/runner/work/waku/waku/node_modules/.pnpm/ps-tree@1.2.0/node_modules/ps-tree/index.js:87:7
        at Stream.a.end (/Users/runner/work/waku/waku/node_modules/.pnpm/event-stream@3.3.4/node_modules/event-stream/index.js:87:5)
        at end (/Users/runner/work/waku/waku/node_modules/.pnpm/map-stream@0.1.0/node_modules/map-stream/index.js:116:39)
        at Stream.stream.end (/Users/runner/work/waku/waku/node_modules/.pnpm/map-stream@0.1.0/node_modules/map-stream/index.js:122:5)
        at drain (/Users/runner/work/waku/waku/node_modules/.pnpm/through@2.3.8/node_modules/through/index.js:34:23)
        at Stream.stream.queue.stream.push (/Users/runner/work/waku/waku/node_modules/.pnpm/through@2.3.8/node_modules/through/index.js:45:5)
        at Stream.<anonymous> (/Users/runner/work/waku/waku/node_modules/.pnpm/split@0.3.3/node_modules/split/index.js:60:10)
        at _end (/Users/runner/work/waku/waku/node_modules/.pnpm/through@2.3.8/node_modules/through/index.js:65:9)
        at Stream.stream.end (/Users/runner/work/waku/waku/node_modules/.pnpm/through@2.3.8/node_modules/through/index.js:74:5)
```

**Comment by ojj1123** [comment]
All tests are passed. Could you review this PR?

File: packages/create-waku/src/index.ts [hunk 3]
```diff
@@ -202,15 +203,33 @@ async function init() {
...
   // 2. and then install dependencies
-
-  console.log(`\nDone. Now run:\n`);
-  console.log(`${bold(green(`cd ${targetDir}`))}`);
-  console.log(`${bold(green(commands.install))}`);
-  console.log(`${bold(green(commands.dev))}`);
   console.log();
+  console.log(`Installing dependencies by running ${commands.install}...`);
+
+  process.chdir(targetDir);
```
Comment (dai-shi) [inline packages/create-waku/src/index.ts:211]: This changes the current working directory and won't go back.

Should we use `cwd` option in `spawn`?

**Review by dai-shi — COMMENTED** [review]
It looks clean!

Let's wait for @himself65 feedback about the spec.

File: packages/create-waku/src/index.ts (hunk 3 above)
Comment (ojj1123) [reply packages/create-waku/src/index.ts:211]: > This changes the current working directory and won't go back.

Because of this, what do you concern with?

Comment (dai-shi) [reply packages/create-waku/src/index.ts:211]: a) I don't like this side-effecty behavior
b) The message at L228 is wrong.

Comment (ojj1123) [reply packages/create-waku/src/index.ts]: 1ad5c5d

I got your thought! check this commit :)

File: e2e/create-waku.spec.ts (hunk 2 above)
Comment (himself65) [reply e2e/create-waku.spec.ts:49]: terminate will stop all the child process realted to the pid, I think the issues was I cannot stop the cp by `cp.kill()` and wait for exit for some reason

**Review by himself65 — APPROVED** [review]

File: e2e/create-waku.spec.ts (hunk 2 above)
Comment (dai-shi) [reply e2e/create-waku.spec.ts:49]: If it's confirmed that it's not necessary, remove it entirely L47-49.

**Review by dai-shi — COMMENTED** [review]
LGTM

File: e2e/create-waku.spec.ts (hunk 2 above)
Comment (ojj1123) [reply e2e/create-waku.spec.ts]: Yeah It is not necessary. 

Check this commit! 1c134df

**Review by dai-shi — APPROVED** [review]
LGTM
Thanks for your contribution!

**Pull request merged**